### PR TITLE
    Extended Learning mode with per VLAN learn mode configuration.

### PR DESCRIPTION
**Pull request by rivasiv**
```
Proposed behaviour is like the following:

|--------------------------------------------------|
| Port learning  | VLAN learning | Result Learning |
|--------------------------------------------------|
| Enable         | Disable       | Port Enable     |
|--------------------------------------------------|
| Enable         | Enable        | Port Enable     |
|--------------------------------------------------|
| Disable        | Disable       | Disable         |
|--------------------------------------------------|
| Disable        | Enable        | VLAN Enable     |
|--------------------------------------------------|
```

Signed-off-by: Rostyslav Ivasiv Rostyslav.Ivasiv@caviumnetworks.com

### CODE REVIEW

**Comment by lguohan** [comment]
not sure I understand the column "Result learning". For the second row, what does the "Port enable" here. Say port 1 is enabled and vlan100 is disabled. Suppose port 1 is also in vlan100, for vlan100 packet comes in port 1, does the ASIC do mac learning or not? Which mac learning setting should we apply to this packet, the setting belongs to the vlan or the port?

**Comment by rivasiv** [comment]
Here we are talking about configuration priorities. Usually more specific configuration takes precedence over more general one. In added design port configuration will have higher priority than configuration over VLAN, let me rephrase the table:

|--------------------------------------------------|-----------------|
| Port learning  | VLAN learning | Learning state  | Applied from    |
|--------------------------------------------------|-----------------|
| Enable         | Disable       | Enable          | Port            |
|--------------------------------------------------|-----------------|
| Enable         | Enable        | Enable          | Port            |
|--------------------------------------------------|-----------------|
| Disable        | Disable       | Disable         | ~               |
|--------------------------------------------------|-----------------|
| Disable        | Enable        | Enable          | VLAN            |
|--------------------------------------------------|-----------------|

For case you described behaviour suppose to be:  The ASIC does MAC learning and uses the port settings.

**Comment by rivasiv** [comment]
Hi @lguohan, What do you think would it have sense?

**Comment by kcudnik** [comment]
Closing, moved to sai 1.0
